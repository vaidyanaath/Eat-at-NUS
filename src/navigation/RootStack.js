import React, { useState, useEffect } from 'react';

// import navigation
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';

// import components
import { ProfileButton } from '../components/buttons/ProfileButton';

// import colors
import { colors } from '../assets/colors';

// import screens
import Landing from '../screens/auth/Landing';
import CustomerSignIn from '../screens/auth/CustomerSignIn';
import CustomerSignUp from '../screens/auth/CustomerSignUp';
import Home from '../screens/main/customer/Home';
import Stall from '../screens/main/customer/Stall';
import Dish from '../screens/main/customer/Dish';

import FoodStallOwnerSignIn from '../screens/auth/FoodStallOwnerSignIn';
import FoodStallOwnerSignUp from '../screens/auth/FoodStallOwnerSignUp';
import StallOwnerHome from '../screens/main/foodStallOwner/StallOwnerHome';
import StallOwnerDish from '../screens/main/foodStallOwner/StallOwnerDish';
import StallOwnerEditDish from '../screens/main/foodStallOwner/StallOwnerEditDish';
import StallOwnerAddDish from '../screens/main/foodStallOwner/StallOwnerAddDish';

// import auth
import { auth } from '../firebase/config';
import { onAuthStateChanged } from 'firebase/auth';

// Import Database
import { ref, onValue } from 'firebase/database';
import { db } from '../firebase/config';

const RootStack = () => {
  const Stack = createNativeStackNavigator();
  const [isSignedIn, setIsSignedIn] = useState(false);
  const [userType, setUserType] = useState('anonymous');

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      if (user) {
        setIsSignedIn(true);
        console.log('Signed in');
        console.log(user.email);
      } else {
        setIsSignedIn(false);
        console.log('Signed out');
      }
    });
    return unsubscribe;
  }, [db]);

  // get user type of entered email
  useEffect(() => {
    if (isSignedIn) {
      const email_key = auth.currentUser.email.replace('.', '%2E');
      const userTypeReference = ref(db, 'userType/' + email_key);
      onValue(userTypeReference, (snapshot) => {
        setUserType(snapshot.val());
        console.log('Fetching user type: ' + userType);
      });
    }
  });

  return (
    <NavigationContainer>
      <Stack.Navigator
        screenOptions={{
          headerStyle: {
            backgroundColor: colors.bg,
            borderBottomWidth: 0,
            height: 120,
            // shadowColor: "transparent",
          },
          headerRight: () => (isSignedIn ? <ProfileButton /> : null),
          headerShadowVisible: false,
          headerTintColor: colors.secondary,
          headerRightContainerStyle: {
            paddingRight: 25,
          },
          headerLeftContainerStyle: {
            paddingLeft: 25,
          },
        }}
      >
        {isSignedIn ? (
          userType === 'customer' ? (
            <>
              <Stack.Screen name="Home" component={Home} options={{ headerShown: false }} />
              <Stack.Screen
                name="Stall"
                component={Stall}
                options={({ route }) => ({ title: route.params.stall.name })}
              />
              <Stack.Screen name="Dish" component={Dish} options={{ title: null }} />
            </>
          ) : (
            <>
              <Stack.Screen
                name="StallOwnerHome"
                component={StallOwnerHome}
                options={{ headerShown: false }}
              />
              <Stack.Screen
                name="StallOwnerDish"
                component={StallOwnerDish}
                options={{ title: null }}
              />
              <Stack.Screen
                name="StallOwnerEditDish"
                component={StallOwnerEditDish}
                options={{ title: null }}
              />
              <Stack.Screen
                name="StallOwnerAddDish"
                component={StallOwnerAddDish}
                options={{ title: null }}
              />
            </>
          )
        ) : (
          <>
            <Stack.Screen
              name="Landing"
              component={Landing}
              options={{ headerShown: false, animationTypeForReplace: isSignedIn ? 'push' : 'pop' }}
            />
            <Stack.Screen
              name="CustomerSignIn"
              component={CustomerSignIn}
              options={{ title: null }}
            />
            <Stack.Screen
              name="CustomerSignUp"
              component={CustomerSignUp}
              options={{ title: null }}
            />
            <Stack.Screen
              name="FoodStallOwnerSignIn"
              component={FoodStallOwnerSignIn}
              options={{ title: null }}
            />
            <Stack.Screen
              name="FoodStallOwnerSignUp"
              component={FoodStallOwnerSignUp}
              options={{ title: null }}
            />
          </>
        )}
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default RootStack;
