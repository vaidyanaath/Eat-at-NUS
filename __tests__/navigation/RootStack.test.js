import React from 'react';
import renderer from 'react-test-renderer';

import RootStack from '../../src/navigation/RootStack';

describe('<RootStack />', () => {
    
    it('Renders correctly', () => {
        const tree = renderer.create(<RootStack />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    
});