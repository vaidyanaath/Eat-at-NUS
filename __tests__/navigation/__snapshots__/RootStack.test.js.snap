// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RootStack /> Renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      forwardedRef={[Function]}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="pop"
      stackPresentation="push"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backgroundColor="#FFFFFF"
        color="#2C365A"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={true}
        hideBackButton={false}
        hideShadow={true}
        largeTitleHideShadow={false}
        title="Landing"
        titleColor="#2C365A"
        topInsetEnabled={false}
        translucent={false}
      >
        <RNSScreenStackHeaderSubview
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          type="right"
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "borderRadius": 20,
                "height": 40,
                "opacity": 1,
                "width": 40,
              }
            }
          >
            <Image
              style={
                Object {
                  "borderRadius": 20,
                  "height": 40,
                  "width": 40,
                }
              }
            />
          </View>
        </RNSScreenStackHeaderSubview>
      </RNSScreenStackHeaderConfig>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "rgb(242, 242, 242)",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "justifyContent": "center",
                "width": "100%",
              },
              Object {
                "alignItems": "center",
                "backgroundColor": "#2C365A",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
                "padding": 0,
              },
            ]
          }
        >
          <Image
            source={1}
            style={
              Object {
                "height": 150,
                "width": 150,
              }
            }
          />
          <Text
            style={
              Array [
                Object {
                  "color": "#2C365A",
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 33,
                },
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "SourceSansPro-Bold",
                  "fontSize": 35,
                  "justifyContent": "flex-end",
                  "marginBottom": 60,
                  "marginTop": 10,
                  "textAlign": "center",
                },
              ]
            }
          >
             Eat@NUS 
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#2C365A",
                  "fontFamily": "SourceSansPro-Regular",
                  "fontSize": 23,
                },
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 30,
                  "marginBottom": 10,
                },
              ]
            }
          >
             I am a . . . 
          </Text>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#ff7f50",
                "borderRadius": 10,
                "elevation": 3,
                "height": 70,
                "justifyContent": "center",
                "marginBottom": 10,
                "marginLeft": 5,
                "marginRight": 5,
                "marginTop": 5,
                "opacity": 1,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.22,
                "shadowRadius": 2.22,
                "width": 280,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#2C365A",
                    "fontFamily": "SourceSansPro-SemiBold",
                    "fontSize": 33,
                  },
                  Object {
                    "color": "#F5F5F5",
                    "fontSize": 25,
                  },
                ]
              }
            >
               Customer 
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#ff7f50",
                "borderRadius": 10,
                "elevation": 3,
                "height": 70,
                "justifyContent": "center",
                "marginBottom": 10,
                "marginLeft": 5,
                "marginRight": 5,
                "marginTop": 5,
                "opacity": 1,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.22,
                "shadowRadius": 2.22,
                "width": 280,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#2C365A",
                    "fontFamily": "SourceSansPro-SemiBold",
                    "fontSize": 33,
                  },
                  Object {
                    "color": "#F5F5F5",
                    "fontSize": 25,
                  },
                ]
              }
            >
               Food Stall Owner 
            </Text>
          </View>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
